{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cb09e6c0-347c-48cb-b5ac-8d85bb883fba",
   "metadata": {},
   "source": [
    "## subset regions of GTF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92e9895c-c9d3-4855-bd0d-a7d84fbe008d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "acaafb78-e50b-4c53-aedc-ebf96286724b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nas/longleaf/rhel8/apps/anaconda/2021.11.ood/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "gtf = pd.read_csv(\"/proj/RNA_lab/Genomic_Indexes/Human/Homo_sapiens.GRCh38.108.chr.gtf\", skiprows=5, header=None, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ec8c8226-32e8-4c16-ac31-af9d9c76bb0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408725</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408726</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408727</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408728</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408729</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3408730 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1            2        3        4  5  6  7  \\\n",
       "0         1  ensembl_havana         gene  1471765  1497848  .  +  .   \n",
       "1         1  ensembl_havana   transcript  1471765  1497848  .  +  .   \n",
       "2         1  ensembl_havana         exon  1471765  1472089  .  +  .   \n",
       "3         1  ensembl_havana          CDS  1471885  1472089  .  +  0   \n",
       "4         1  ensembl_havana  start_codon  1471885  1471887  .  +  0   \n",
       "...      ..             ...          ...      ...      ... .. .. ..   \n",
       "3408725  MT           insdc   transcript    15888    15953  .  +  .   \n",
       "3408726  MT           insdc         exon    15888    15953  .  +  .   \n",
       "3408727  MT           insdc         gene    15956    16023  .  -  .   \n",
       "3408728  MT           insdc   transcript    15956    16023  .  -  .   \n",
       "3408729  MT           insdc         exon    15956    16023  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...                                                    ...  \n",
       "3408725  gene_id \"ENSG00000210195\"; gene_version \"2\"; t...  \n",
       "3408726  gene_id \"ENSG00000210195\"; gene_version \"2\"; t...  \n",
       "3408727  gene_id \"ENSG00000210196\"; gene_version \"2\"; g...  \n",
       "3408728  gene_id \"ENSG00000210196\"; gene_version \"2\"; t...  \n",
       "3408729  gene_id \"ENSG00000210196\"; gene_version \"2\"; t...  \n",
       "\n",
       "[3408730 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d3d70463-1ab6-4a10-a839-b25764b5faea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1477274</td>\n",
       "      <td>1477350</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408532</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>14964716</td>\n",
       "      <td>14964718</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408533</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>15064745</td>\n",
       "      <td>15065000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408534</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>15014344</td>\n",
       "      <td>15014466</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408535</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>14968193</td>\n",
       "      <td>14968526</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408536</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>14961235</td>\n",
       "      <td>14964715</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1180023 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1                2         3         4  5  6  7  \\\n",
       "1         1  ensembl_havana       transcript   1471765   1497848  .  +  .   \n",
       "2         1  ensembl_havana             exon   1471765   1472089  .  +  .   \n",
       "3         1  ensembl_havana              CDS   1471885   1472089  .  +  0   \n",
       "4         1  ensembl_havana      start_codon   1471885   1471887  .  +  0   \n",
       "5         1  ensembl_havana             exon   1477274   1477350  .  +  .   \n",
       "...      ..             ...              ...       ...       ... .. .. ..   \n",
       "3408532  21          havana       stop_codon  14964716  14964718  .  -  0   \n",
       "3408533  21          havana   five_prime_utr  15064745  15065000  .  -  .   \n",
       "3408534  21          havana   five_prime_utr  15014344  15014466  .  -  .   \n",
       "3408535  21          havana   five_prime_utr  14968193  14968526  .  -  .   \n",
       "3408536  21          havana  three_prime_utr  14961235  14964715  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "1        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "5        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...                                                    ...  \n",
       "3408532  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408533  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408534  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408535  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408536  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "\n",
       "[1180023 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CCDS is the consensus of coding sequences from NCBI \n",
    "ccds = gtf[gtf[8].str.contains(\"CCDS\")]\n",
    "ccds "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "57fd2b32-4963-4b69-b6ed-69c9d4f40969",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1477274</td>\n",
       "      <td>1477350</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1478644</td>\n",
       "      <td>1478745</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1479049</td>\n",
       "      <td>1479108</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1480867</td>\n",
       "      <td>1480936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408032</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>46126503</td>\n",
       "      <td>46126541</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000142173\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408034</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>46128903</td>\n",
       "      <td>46128925</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000142173\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408506</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14964719</td>\n",
       "      <td>14968192</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408518</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14964719</td>\n",
       "      <td>14968192</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408530</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14964719</td>\n",
       "      <td>14968192</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>442037 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1    2         3         4  5  6  7  \\\n",
       "3         1  ensembl_havana  CDS   1471885   1472089  .  +  0   \n",
       "6         1  ensembl_havana  CDS   1477274   1477350  .  +  2   \n",
       "8         1  ensembl_havana  CDS   1478644   1478745  .  +  0   \n",
       "10        1  ensembl_havana  CDS   1479049   1479108  .  +  0   \n",
       "12        1  ensembl_havana  CDS   1480867   1480936  .  +  0   \n",
       "...      ..             ...  ...       ...       ... .. .. ..   \n",
       "3408032  21          havana  CDS  46126503  46126541  .  +  2   \n",
       "3408034  21          havana  CDS  46128903  46128925  .  +  2   \n",
       "3408506  21          havana  CDS  14964719  14968192  .  -  0   \n",
       "3408518  21  ensembl_havana  CDS  14964719  14968192  .  -  0   \n",
       "3408530  21          havana  CDS  14964719  14968192  .  -  0   \n",
       "\n",
       "                                                         8  \n",
       "3        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "6        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "8        gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "10       gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "12       gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...                                                    ...  \n",
       "3408032  gene_id \"ENSG00000142173\"; gene_version \"17\"; ...  \n",
       "3408034  gene_id \"ENSG00000142173\"; gene_version \"17\"; ...  \n",
       "3408506  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408518  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408530  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "\n",
       "[442037 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds = ccds[ccds[2] == \"CDS\"]\n",
    "cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3f93b27d-b2f4-4f6d-bd3a-d11bf194a8f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>1495818</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>3433812</td>\n",
       "      <td>3435421</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>3433812</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2405765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>2403974</td>\n",
       "      <td>2405765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3407969</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>46132553</td>\n",
       "      <td>46132848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142173\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408036</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>46128929</td>\n",
       "      <td>46129808</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142173\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408512</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>14961235</td>\n",
       "      <td>14964715</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408525</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>14961235</td>\n",
       "      <td>14964715</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408536</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>14961235</td>\n",
       "      <td>14964715</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47686 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1                2         3         4  5  6  7  \\\n",
       "37        1  ensembl_havana  three_prime_utr   1495818   1497848  .  +  .   \n",
       "198       1          havana  three_prime_utr   3433812   3435421  .  +  .   \n",
       "274       1  ensembl_havana  three_prime_utr   3433812   3438621  .  +  .   \n",
       "362       1  ensembl_havana  three_prime_utr   2403964   2405765  .  -  .   \n",
       "379       1          havana  three_prime_utr   2403974   2405765  .  -  .   \n",
       "...      ..             ...              ...       ...       ... .. .. ..   \n",
       "3407969  21  ensembl_havana  three_prime_utr  46132553  46132848  .  +  .   \n",
       "3408036  21          havana  three_prime_utr  46128929  46129808  .  +  .   \n",
       "3408512  21          havana  three_prime_utr  14961235  14964715  .  -  .   \n",
       "3408525  21  ensembl_havana  three_prime_utr  14961235  14964715  .  -  .   \n",
       "3408536  21          havana  three_prime_utr  14961235  14964715  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "37       gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "198      gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "274      gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "362      gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "379      gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "...                                                    ...  \n",
       "3407969  gene_id \"ENSG00000142173\"; gene_version \"17\"; ...  \n",
       "3408036  gene_id \"ENSG00000142173\"; gene_version \"17\"; ...  \n",
       "3408512  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408525  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408536  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "\n",
       "[47686 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr3 = ccds[ccds[2] == \"three_prime_utr\"]\n",
    "utr3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9d40dbc0-78bc-4205-adb4-f63dbf06769e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1471884</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>3069197</td>\n",
       "      <td>3069259</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>3069203</td>\n",
       "      <td>3069259</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>2412503</td>\n",
       "      <td>2412564</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>2412503</td>\n",
       "      <td>2412564</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408523</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>15014344</td>\n",
       "      <td>15014466</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408524</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>14968193</td>\n",
       "      <td>14968526</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408533</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>15064745</td>\n",
       "      <td>15065000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408534</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>15014344</td>\n",
       "      <td>15014466</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408535</th>\n",
       "      <td>21</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>14968193</td>\n",
       "      <td>14968526</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180530\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76514 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1               2         3         4  5  6  7  \\\n",
       "36        1  ensembl_havana  five_prime_utr   1471765   1471884  .  +  .   \n",
       "197       1          havana  five_prime_utr   3069197   3069259  .  +  .   \n",
       "273       1  ensembl_havana  five_prime_utr   3069203   3069259  .  +  .   \n",
       "361       1  ensembl_havana  five_prime_utr   2412503   2412564  .  -  .   \n",
       "378       1          havana  five_prime_utr   2412503   2412564  .  -  .   \n",
       "...      ..             ...             ...       ...       ... .. .. ..   \n",
       "3408523  21  ensembl_havana  five_prime_utr  15014344  15014466  .  -  .   \n",
       "3408524  21  ensembl_havana  five_prime_utr  14968193  14968526  .  -  .   \n",
       "3408533  21          havana  five_prime_utr  15064745  15065000  .  -  .   \n",
       "3408534  21          havana  five_prime_utr  15014344  15014466  .  -  .   \n",
       "3408535  21          havana  five_prime_utr  14968193  14968526  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "36       gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "197      gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "273      gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "361      gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "378      gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "...                                                    ...  \n",
       "3408523  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408524  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408533  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408534  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "3408535  gene_id \"ENSG00000180530\"; gene_version \"11\"; ...  \n",
       "\n",
       "[76514 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr5 = ccds[ccds[2] == \"five_prime_utr\"]\n",
    "utr5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2bd8accb-d4c1-4b0d-9a55-13b081e3ec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# comment out when done saving \n",
    "# cds.to_csv(\"/nas/longleaf/home/askav/rna/autoregulation/g4/CDS.subset.gtf\")\n",
    "# utr5.to_csv(\"/nas/longleaf/home/askav/rna/autoregulation/g4/UTR3.subset.gtf\")\n",
    "# utr3.to_csv(\"/nas/longleaf/home/askav/rna/autoregulation/g4/UTR5.subset.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "061da263-c0de-4679-b42c-83121055134f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
